docs/architecture/adr-004-genesis-modes.md

```diff
@@ -0,0 +1,295 @@
+# ADR 004: Genesis Modes
+
+## Changelog
+
+- Sept 3, 2019: Initial draft
+
+## Context
+
+Each module that implements the `AppModuleGenesis` interface is required to fulfill
+the `InitGenesis` contract. Typically, the business logic for any module's `InitGenesis`
+implementation revolves around setting values and parameters provided to that module
+from a genesis application state.
+
+Example from `x/auth`:
+
+```go
+func InitGenesis(ctx sdk.Context, ak AccountKeeper, data GenesisState) {
+  ak.SetParams(ctx, data.Params)
+  data.Accounts = SanitizeGenesisAccounts(data.Accounts)
+
+  for _, a := range data.Accounts {
+    acc := ak.NewAccount(ctx, a)
+    ak.SetAccount(ctx, acc)
+  }
+}
+```
+
+However, some modules, in addition to setting parameters and values from genesis state,
+also _modify_ the genesis state implicitly when certain values are not provided.
+At the time of this ADR's draft, this only happens when module accounts are not
+provided in the genesis application state.
+
+Example from `x/gov`:
+
+```go
+func InitGenesis(ctx sdk.Context, k Keeper, supplyKeeper types.SupplyKeeper, data GenesisState) {
+  // ...
+
+  // add coins if not provided on genesis
+  if moduleAcc.GetCoins().IsZero() {
+    if err := moduleAcc.SetCoins(totalDeposits); err != nil {
+      panic(err)
+    }
+  
+    supplyKeeper.SetModuleAccount(ctx, moduleAcc)
+  }
+}
+```
+
+This kind of business logic in `InitGenesis` is not desirable because it modifies
+genesis state without explicitly doing so and as a result can lead to bugs, irregularities,
+and a confusing UX if the user is not aware of these implicit modifications.
+
+## Decision
```

Comment:
I have not read through the entirety of the `Decision`, but from the looks of it, it seems quite verbose and overly complicated. In addition, I have no idea why we're modifying BaseApp. We should really avoid doing so if possible.

From the chat I had with @rigelrozanski the solution here should be pretty simple:

- Update the `AppModuleGenesis` interface to also now include `PopulateGenesis` with very clear documentation on what this does.
- Create a  generic `populate-genesis` command that goes through each module in a genesis file and calls `PopulateGenesis` on each module and returns the resulting genesis state.

Users, therefore, have the ability to "opt-in" into this functionality. In addition, we have the ability to this programmatically (via looping and calling PopulateGenesis) which will be useful for simulation and exporting and manually (via a command). Thoughts?